### ISSUE TITLE
Weird broken loop with 308 Permanent Redirect on php_fastcgi

### ISSUE THREAD

**Issue by SirLouen** [caddyserver/caddy#6496]
I've noted that `php_fastcgi` directive has a default behaviour that can completely break the Let's Encrypt retrieval process

I have this test domain: fedora.mcamargo.es

Configured with this Caddyfile
```
fedora.mcamargo.es {
        root * /usr/share/caddy/fedora
        file_server
        php_fastcgi unix//run/php/php-fpm.sock {
                root /var/www/html
        }
}
```

I'm using a dockerized Caddy (caddy:latest 2.8.4) with an another dockerized PHP-FPM server
At some point I deleted the volume where certs were stored. 

As we can see here this is the response:

```
curl http://fedora.mcamargo.es -v
* Host fedora.mcamargo.es:80 was resolved.
* IPv6: (none)
* IPv4: 135.181.80.106
*   Trying 135.181.80.106:80...
* Connected to fedora.mcamargo.es (135.181.80.106) port 80
> GET / HTTP/1.1
> Host: fedora.mcamargo.es
> User-Agent: curl/8.7.1
> Accept: */*
>
* Request completely sent off
< HTTP/1.1 308 Permanent Redirect
< Connection: close
< Location: https://fedora.mcamargo.es/
< Server: Caddy
< Date: Mon, 05 Aug 2024 10:47:13 GMT
< Content-Length: 0
<
* Closing connection
```

There is a permanent 308 redirect that inhibits the recreation of the new certificate, just as we can see in the letsdebug error:

https://letsdebug.net/fedora.mcamargo.es/2155062

I'm going to research how I can stop that 308 redirect to HTTPS to re-issue the certificates, but the system is broken as-is and I have not idea for now how to sort this out.

I'm not sure if this issue is related somehow: https://github.com/caddyserver/caddy/issues/5660 as I cannot relate to the content of it, but maybe there is some relationship

If I change the hostname in the Caddyfile to 

```
http://fedora.mcamargo.es {
        root * /usr/share/caddy/fedora
        file_server
        php_fastcgi unix//run/php/php-fpm.sock {
                root /var/www/html
        }
}
```

As we can see here, it's fully compliant again.

https://letsdebug.net/fedora.mcamargo.es/2155077

And the site works well (without HTTPS)

That pesky 308 is driving me nuts :)

**Comment by mholt** [comment]
And you're saying that this doesn't happen without php_fastcgi?

**Comment by SirLouen** [comment]
> And you're saying that this doesn't happen without php_fastcgi?

I've run out of ideas. Recently I discovered that this doesnt doent use certbot but a self signing software, 

And logs are not saying anything to me

```
l":"info","ts":1722856744.7359416,"logger":"tls.obtain","msg":"lock acquired","identifier":"fedora.mcamargo.es"}
{"level":"info","ts":1722856744.7362647,"logger":"tls.obtain","msg":"obtaining certificate","identifier":"fedora.mcamargo.es"}
{"level":"info","ts":1722856744.7378972,"logger":"http","msg":"using ACME account","account_id":"https://acme-v02.api.letsencrypt.org/acme/acct/1873678947","account_contact":[]}
{"level":"info","ts":1722856744.7379506,"logger":"http","msg":"waiting on internal rate limiter","identifiers":["fedora.mcamargo.es"],"ca":"https://acme-v02.api.letsencrypt.org/directory","account":""}
{"level":"info","ts":1722856744.7380552,"logger":"http","msg":"done waiting on internal rate limiter","identifiers":["fedora.mcamargo.es"],"ca":"https://acme-v02.api.letsencrypt.org/directory","account":""}
{"level":"error","ts":1722856745.4084947,"logger":"tls.obtain","msg":"could not get certificate from issuer","identifier":"fedora.mcamargo.es","issuer":"acme-v02.api.letsencrypt.org-directory","error":"HTTP 400 urn:ietf:params:acme:error:malformed - JWS verification error"}
{"level":"error","ts":1722856745.4086013,"logger":"tls.obtain","msg":"will retry","error":"[fedora.mcamargo.es] Obtain: [fedora.mcamargo.es] creating new order: attempt 1: https://acme-v02.api.letsencrypt.org/acme/new-order: HTTP 400 urn:ietf:params:acme:error:malformed - JWS verification error (ca=https://acme-v02.api.letsencrypt.org/directory)","attempt":1,"retrying_in":60,"elapsed":0.672511664,"max_duration":2592000}
```

I tried with 

```
http://fedora.mcamargo.es, fedora.mcamargo.es {
...
}

And despite with these hostnames, I can get the right "All OK" in all 3 letsdebug tests:
https://letsdebug.net/fedora.mcamargo.es/2155278
https://letsdebug.net/fedora.mcamargo.es/2155281
https://letsdebug.net/fedora.mcamargo.es/2155284

Same error above pops. I'm 100% clueless right now.

Can I get like more verbosity?

**Comment by mholt** [comment]
But does it work when you simply remove `php_fastcgi`? You mention that in the topic title but then your test case doesn't do anything with `php_fastcgi`, instead it adds `http://` which of course disables HTTPS. I'm trying to figure out how php_fastcgi is relevant here.

**Comment by SirLouen** [comment]
> But does it work when you simply remove `php_fastcgi`? You mention that in the topic title but then your test case doesn't do anything with `php_fastcgi`, instead it adds `http://` which of course disables HTTPS. I'm trying to figure out how php_fastcgi is relevant here.

I've commented out the php_fascgi part, and I've realized that php_fastcgi is irrelevant. I saw in the docs that there was a 408 Redirect but its irrelevant. The 408 seemed to be the HTTPS 408 redirect from another redir rule, and I got confused with the 408 from the php_fastcgi one. 

So it appears that the issue here is beyond php_faastcgi, and I'm completely unaware of what could be provoking it at this moment. I'm just testing random things.

**Comment by mholt** [comment]
Ok, so, start with a new (virtual) machine.

Update your A/AAAA records so they both point to it.

Set this to be your Caddyfile:

```
{
    debug
}

fedora.mcamargo.es {
        respond "Hello world"
}
```

and run `caddy run > caddy.log 2>&1` and post the contents of the file as well as the results of `curl -v "https://fedora.mcamargo.es"` after a certificate has been provisioned.

Also please post the output of `caddy build-info`.

**Comment by francislavoie** [comment]
I don't think there's any problem here. I think you're just getting misled by Let's Debug. Caddy definitely handles ACME HTTP challenges correctly. If the path requested doesn't map to an HTTP challenge that exists in Caddy's storage, then Caddy will let the request fall through to default HTTP handling, resulting in a HTTP->HTTPS redirect. That's normal.

**Comment by SirLouen** [comment]
> Ok, so, start with a new (virtual) machine.

Let's see

docker-compose.yml

```
caddyv2:
    image: caddy:latest
    container_name: caddyv2
    restart: unless-stopped
    ports:
      - 80:80
      - 443:443
    volumes:
      - $PWD/caddyv2:/etc/caddy
```

`caddyv2/Caddyfile`

```
{
    debug
}

fedora.mcamargo.es {
        respond "Hello world"
}
```

And yes, I can confirm that Caddy is a working software as expected

![image](https://github.com/user-attachments/assets/415564a6-10e3-4203-8668-05f94e678f96)

The question is: why is not working under the conditions of the other image? 

They are both `caddy:latest` with the provided config Caddyfiles

@francislavoie mentioned something about storage.

The other has this volumes in docker compose:

```
volumes:
      - caddy_data:/data/caddy
      - caddy_config:/config
      - $PWD/volumes/wordpress:/usr/share/wordpress
      - phpmyadmin_data:/usr/share/phpmyadmin
      - $PWD/caddy:/etc/caddy
      - php_socket:/run/php
```

Being:

1. The first two, I believe two classic ones with regular docker volumes
2. The next two, two binds to the FPM content directories for the php_fastcgi socket. There are more sites apart from fedora.mcamargo.es that I put as an example all failing with identical config as the one provided above.
3. The fifth, is the Caddyfile folder
4. And the last is the shared socket folder between FPM and Caddy.

Yesterday everything was working well.

The only "big" change I did was switching from regular mounts to docker volumes on caddy_data and caddy_config and maybe something else I cannot remember.

I'm testing things like mad, but at first glance it doesnt make any sense.

This are the initial logs 

```
{"level":"info","ts":1722861953.3492475,"logger":"http","msg":"enabling HTTP/3 listener","addr":":443"}
{"level":"info","ts":1722861953.3494942,"msg":"failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 7168 kiB, got: 416 kiB). See https://github.com/quic-go/quic-go/wiki/UDP-Buffer-Sizes for details."}
{"level":"debug","ts":1722861953.3498461,"logger":"http","msg":"starting server loop","address":"[::]:443","tls":true,"http3":true}
{"level":"info","ts":1722861953.3498998,"logger":"http.log","msg":"server running","name":"srv0","protocols":["h1","h2","h3"]}
{"level":"debug","ts":1722861953.3500166,"logger":"http","msg":"starting server loop","address":"[::]:80","tls":false,"http3":false}
{"level":"info","ts":1722861953.3500504,"logger":"http.log","msg":"server running","name":"srv1","protocols":["h1","h2","h3"]}
{"level":"info","ts":1722861953.3500743,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["fedora.mcamargo.es"]}
{"level":"info","ts":1722861953.35041,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
{"level":"info","ts":1722861953.3504834,"msg":"serving initial configuration"}
{"level":"info","ts":1722861953.3509104,"logger":"tls.obtain","msg":"acquiring lock","identifier":"fedora.mcamargo.es"}
{"level":"info","ts":1722861953.372934,"logger":"tls","msg":"storage cleaning happened too recently; skipping for now","storage":"FileStorage:/data/caddy","instance":"7bec2d31-f920-4348-b44a-fd5d01d128eb","try_again":1722948353.3729286,"try_again_in":86399.999998973}
{"level":"info","ts":1722861953.3730545,"logger":"tls","msg":"finished cleaning storage units"}
{"level":"info","ts":1722861953.3731596,"logger":"tls.obtain","msg":"lock acquired","identifier":"fedora.mcamargo.es"}
{"level":"info","ts":1722861953.3735702,"logger":"tls.obtain","msg":"obtaining certificate","identifier":"fedora.mcamargo.es"}
```

Not sure if they could be relevant.

It says "storage clearning happened to recently". I'm taking shots in the dark.

**Comment by SirLouen** [comment]
@mholt 
Ok something new discovered
After doing everything above, I thought that I never changed the volumes back so I switch back to:

```
   volumes:
      - $PWD/caddy/data:/data/caddy
      - $PWD/caddy/config:/config
      - $PWD/volumes/wordpress:/usr/share/wordpress
      - phpmyadmin_data:/usr/share/phpmyadmin
      - $PWD/caddy:/etc/caddy
      - php_socket:/run/php
  ```
  
  Instead of using `caddy_config` and `caddy_data` I simply used two directories within the directory where the docker-compose.yml, and it worked (with another domain, because fedora.mcamargo.es has been rated limited (luckily I saved the certificates in the last trial)

So definitely I can say that the culprit is with those volumes... Very strange I have to admit and definitely impossible to track down with the logs (I just remembered some minutes ago that I changed that yesterday while I was doing something else, and completely forgot about that change until now).

I wonder whats going on with such volumes. They are very basic `driver:local` docker volumes, nothing special about them, and working for every other scenario, like the phpmyadmin files, etc.. 

Why the certificate part could be having trouble?

**Comment by francislavoie** [comment]
This is our recommended Docker Compose setup: https://caddyserver.com/docs/running#docker-compose

**Comment by SirLouen** [comment]
> This is our recommended Docker Compose setup: https://caddyserver.com/docs/running#docker-compose
> 
> I think the problem is that you're nesting volumes. You should not do that. Don't put a volume in another one. Separate them in different directories.

I'm not nesting volumes

In fact, that config is what its provoking the issues I described above. 

For some reason switching from caddy_data volume to $PWD/caddy/data mount point solved the issue, but it leaves me seriously uneasy

**Comment by erostefano** [comment]
@SirLouen I have the same issue. How exactly did you fix it? This is my docker-compose:

`version: '3'

networks:
  reverse-proxy:
    external: true

services:
  caddy:
    image: caddy:latest
    container_name: caddy
    restart: unless-stopped
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - ./vol/Caddyfile:/etc/caddy/Caddyfile
      - ./vol/data:/data
      - ./vol/config:/config
    networks:
      - reverse-proxy
`

**Comment by SirLouen** [comment]
> @SirLouen I have the same issue. How exactly did you fix it? This is my docker-compose:
> 
> ` version: '3'
> 
> networks: reverse-proxy: external: true
> 
> services: caddy: image: caddy:latest container_name: caddy restart: unless-stopped ports: - '80:80' - '443:443' volumes: - ./vol/Caddyfile:/etc/caddy/Caddyfile - ./vol/data:/data - ./vol/config:/config networks: - reverse-proxy `

I had both caddy config and caddy data in a Volume and I moved them into a directory like you have and it was solved mysteriously. 

There is clearly a malfunction in the system, but @francislavoie decided to close it because, unfortunately, it's not easy reproduceable and even less tractable because debug logs are not saying anything in particular. 

I was about to drop caddy just because of this issue because it was driving me nuts, but it was solved randomly when i switched, as i say, from volumes to regular directories. 

Did you initiate the docker composer originally with volumes?

I always thought, that maybe there is like a hard bind where if you select volumes or folders, and then you switch after the first deployment of the files, the system will induce this error.
